tuta-sdk/rust/src/rest_error.rs

```diff
@@ -0,0 +1,432 @@
+        use PreconditionFailedReason::*;
+
+        // Check and return if the string matches a value of `UnsubscribeFailureReason`
+        match UnsubscribeFailureReason::from_str(s) {
+            Ok(reason) => return Ok(UnsubscribeFailure(reason)),
+            Err(_) => {}
+        }
+
+        match BookingFailureReason::from_str(s) {
+            Ok(reason) => return Ok(BookingFailure(reason)),
+            Err(_) => {}
+        }
+
+        match DomainFailureReason::from_str(s) {
+            Ok(reason) => return Ok(DomainFailure(reason)),
+            Err(_) => {}
+        }
+
+        match CustomDomainFailureReason::from_str(s) {
+            Ok(reason) => return Ok(CustomDomainFailure(reason)),
+            Err(_) => {}
+        }
+
+        match TemplateGroupFailureReason::from_str(s) {
+            Ok(reason) => return Ok(TemplateGroupFailure(reason)),
+            Err(_) => {}
+        }
+
+        match UsageTestFailureReason::from_str(s) {
+            Ok(reason) => return Ok(UsageTestFailure(reason)),
+            Err(_) => {}
+        }
+
+        return match s {
+            "lock.locked" => Ok(FailureLocked),
+            "mailaddressaliasservice.group_disabled" => Ok(FailureUserDisabled),
+            "outofoffice.not_available_on_current_plan" => Ok(FailureUpgradeRequired),
+            _ => Err(ParseFailureError)
+        };
+    }
+}
+
+
+/// The possible error responses from the server
+#[derive(Error, Debug, uniffi::Error, Eq, PartialEq)]
+pub enum NetworkError {
+    #[error("Connection lost")]
+    ConnectionError,
+    #[error("Bad request")]
+    BadRequestError,
+    #[error("Not authenticated")]
+    NotAuthenticatedError,
+    #[error("Not Authorized")]
+    NotAuthorizedError,
+    #[error("Not Found")]
+    NotFoundError,
+    #[error("Method Not Allowed")]
+    MethodNotAllowedError,
+    #[error("Request Timeout")]
+    RequestTimeoutError,
+    #[error("Precondition Failed")]
+    PreconditionFailedError(Option<PreconditionFailedReason>),
+    #[error("Locked")]
+    LockedError,
+    #[error("Too Many Requests")]
+    TooManyRequestsError,
+    #[error("Session Expired")]
+    SessionExpiredError,
+    #[error("Access Deactivated")]
+    AccessDeactivatedError,
+    /// Happens to external users exclusively, due to password changes
+    #[error("Access Expired")]
+    AccessExpiredError,
+    #[error("Access Blocked")]
+    AccessBlockedError,
+    #[error("Invalid Data")]
+    InvalidDataError,
+    #[error("Invalid Software Version")]
+    InvalidSoftwareVersionError,
+    #[error("Limit Reached")]
+    LimitReachedError,
+    #[error("Internal Server")]
+    InternalServerError,
+    #[error("Bad gateway")]
+    BadGatewayError,
+    #[error("Service Unavailable")]
+    ServiceUnavailableError,
+    #[error("Insufficient Storage")]
+    InsufficientStorageError,
+    #[error("Resource")]
+    ResourceError,
+    #[error("Payload Too Large")]
+    PayloadTooLargeError,
+}
+
+impl NetworkError {
```

Comment:
I am not sure that network error is the best name, perhaps HttpError?

Reply:
That is true; It is too vague of a name. Your suggestion sounds good so I have change it to that.